url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/layout/hierarchical/stage/mxCoordinateAssignment.java

```diff
@@ -0,0 +1,1569 @@
+					rankValue + 1);
+			int medianPreviousLevel = medianXValue(previousLayerConnectedCells,
+					rankValue - 1);
+
+			int numConnectedNeighbours = numNextLayerConnected
+					+ numPreviousLayerConnected;
+			int currentPosition = cell.getGeneralPurposeVariable(rankValue);
+			double cellMedian = currentPosition;
+
+			if (numConnectedNeighbours > 0)
+			{
+				cellMedian = (medianNextLevel * numNextLayerConnected + medianPreviousLevel
+						* numPreviousLayerConnected)
+						/ numConnectedNeighbours;
+			}
+
+			// Flag storing whether or not position has changed
+			boolean positionChanged = false;
+
+			if (cellMedian < currentPosition - tolerance)
+			{
+				if (rankIndex == 0)
+				{
+					cell.setGeneralPurposeVariable(rankValue, (int) cellMedian);
+					positionChanged = true;
+				}
+				else
+				{
+					mxGraphAbstractHierarchyCell leftCell = rank[rankValue][rankIndex - 1];
+					int leftLimit = leftCell
+							.getGeneralPurposeVariable(rankValue);
+					leftLimit = leftLimit + (int) leftCell.width / 2
+							+ (int) intraCellSpacing + (int) cell.width / 2;
+
+					if (leftLimit < cellMedian)
+					{
+						cell.setGeneralPurposeVariable(rankValue,
+								(int) cellMedian);
+						positionChanged = true;
+					}
+					else if (leftLimit < cell
+							.getGeneralPurposeVariable(rankValue) - tolerance)
+					{
+						cell.setGeneralPurposeVariable(rankValue, leftLimit);
+						positionChanged = true;
+					}
+				}
+			}
+			else if (cellMedian > currentPosition + tolerance)
+			{
+				int rankSize = rank[rankValue].length;
+
+				if (rankIndex == rankSize - 1)
+				{
+					cell.setGeneralPurposeVariable(rankValue, (int) cellMedian);
+					positionChanged = true;
+				}
+				else
+				{
+					mxGraphAbstractHierarchyCell rightCell = rank[rankValue][rankIndex + 1];
+					int rightLimit = rightCell
+							.getGeneralPurposeVariable(rankValue);
+					rightLimit = rightLimit - (int) rightCell.width / 2
+							- (int) intraCellSpacing - (int) cell.width / 2;
+
+					if (rightLimit > cellMedian)
+					{
+						cell.setGeneralPurposeVariable(rankValue,
+								(int) cellMedian);
+						positionChanged = true;
+					}
+					else if (rightLimit > cell
+							.getGeneralPurposeVariable(rankValue) + tolerance)
+					{
+						cell.setGeneralPurposeVariable(rankValue, rightLimit);
+						positionChanged = true;
+					}
+				}
+			}
+
+			if (positionChanged)
+			{
+				// Add connected nodes to map and list
+				for (int i = 0; i < nextLayerConnectedCells.length; i++)
+				{
+					mxGraphAbstractHierarchyCell connectedCell = (mxGraphAbstractHierarchyCell) nextLayerConnectedCells[i];
+					WeightedCellSorter connectedCellWrapper = map
+							.get(connectedCell);
+
+					if (connectedCellWrapper != null)
+					{
+						if (connectedCellWrapper.visited == false)
+						{
+							connectedCellWrapper.visited = true;
+							nodeList.add(connectedCellWrapper);
+						}
+					}
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 403:417, 420:434.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.